vector_typet::size() is now a constant_exprt

kroening opened the pull request:
The codebase assumes basically globally that vector_typet::size() is a
constant_exprt.  This is now enforced by the signature.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
vector_typet::size() is now a constant_exprt

The codebase assumes basically globally that vector_typet::size() is a
constant_exprt.  This is now enforced by the signature.

tautschnig commented on src/util/std_types.cpp:
```diff
@@ -283,3 +283,19 @@ bool is_constant_or_has_constant_components(
+  : type_with_subtypet(ID_vector, _subtype)
+{
+  size() = _size;
+}
```
What's the rationale to move these to the .cpp file?

kroening replied:
`constant_exprt` isn't complete -- there's a cyclic dependency between std_types.h and std_expr.h.

tautschnig reviewed (approved):
I'm pretty sure this requires further changes in the code base, but of course the stricter type enforcement is good.

kroening commented:
@tautschnig The problem is that there's a period of time in both front-ends where the expression isn't yet turned into a constant. I might have to introduce s.th. like `unfinished_vector_typet` or the like.

kroening committed:
C/C++ now have frontend_vector type

This enables distinguishing the case of a size that is an arbitary
expression from the case where the size is required to be constant.

kroening commented:
Added ID_frontend_vector, similar to ID_frontend_pointer, for the type that's yet to be finished.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 41e491392d8f97b8ffd4a4aefa6bed0f1f62b815).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99471857
